dali/pipeline/data/dynamic_tensor_view.h

```diff
@@ -0,0 +1,623 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_PIPELINE_DATA_DYNAMIC_TENSOR_VIEW_H_
+#define DALI_PIPELINE_DATA_DYNAMIC_TENSOR_VIEW_H_
+
+#include <stdexcept>
+#include <type_traits>
+#include <utility>
+#include <vector>
+#include "dali/core/tensor_shape.h"
+#include "dali/core/tensor_view.h"
+#include "dali/pipeline/data/types.h"
+
+namespace dali {
+
+/**
+ * @brief Tag type indicating that the TensorView is keeping the type information in runtime.
+ *
+ * It can be converted to the statically-typed TensorView using `view<T, ndim>()` function
+ * or by `TensorView::to_static_type<T, ndim>()` member function.
+ */
+struct DynamicType {
+  DynamicType() = delete;
+};
+
+template <typename Backend, typename DataType, int ndim = DynamicDimensions>
+struct DynamicTensorViewBase {
+  static_assert(std::is_same<std::remove_const_t<DataType>, void>::value,
+                "The underlying type must either be void or const void");
+  using element_type = DataType;
+  int dim() const {
+    return shape.sample_dim();
+  }
+  DALIDataType type() const {
+    return type_id;
+  }
+
+  ptrdiff_t num_elements() const {
+    return shape.num_elements();
+  }
+
+  /**
+   * @brief Utility to calculate pointer to element at given coordinates
+   */
+  template <typename... Indices>
+  DataType *operator()(int64_t idx0, Indices &&...idx) const {
+    return data + CalcOffset(shape, std::array<ptrdiff_t, sizeof...(Indices) + 1>{
+                                        idx0, (ptrdiff_t{idx})...});
+  }
+
```

Comment:
This can't work. Your `DataType` is void, so compiling this function would result in `void*` arithmetic.